Add Item predicate for advancement triggers (Item conditions 2/3)

Goldorion opened the pull request:
This is one of many PRs trying to achieve the objective of adding as a standalone output block the item condition that will be used by all item triggers.

This PR adds the item predicate block. Existing JSON triggers that don't support yet this new feature (item enchanted and item damaged) will be updated after this PR is merged as they are more complex to convert due to the necessity of new blocks.

- [x] Merge #5828
- [x] Implementing working item predicate features
- [x] Add tests
- [x] Fix tests :(



-----
Changelog:
- Item consumed now support item predicate block
- Added new advancement item predicate component: Damage component

Goldorion committed:
Add new category + other changes

- Advancement GUI now uses its own XML toolbox file
- Move entry blocks to this new category
- Groups created in the first PR are no longer useful

Goldorion committed:
Any item of output block

Goldorion committed:
Fix blocks not being added to the category

Goldorion committed:
Rename toolbox's file to uniformize names

Goldorion committed:
Fix Any item of not having the item selectors by default

Goldorion committed:
Add an Item condition block

Goldorion committed:
Remove changes added with #5828

Goldorion committed:
I forgot some elements

Goldorion committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into item-entry-related-stuff-advancements

Goldorion committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into item-entry-related-stuff-advancements

# Conflicts:
#	plugins/mcreator-localization/lang/texts.properties

Goldorion committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into item-entry-related-stuff-advancements

# Conflicts:
#	plugins/mcreator-localization/lang/texts.properties

Goldorion committed:
Implement full Item condition features

- Add a full item condition with item, count (stacksize) and components list
- Add a simple item coniditon (item only)
- Add the Damage/durability component with 3 options
- Add a new Item components category
- Group blocks together

Goldorion committed:
Add using_item trigger to use item condition blocks

Goldorion committed:
Implement tests

Goldorion committed:
Default value for item_condition mutator

Goldorion committed:
Fix tests + better item visual for item condition

Goldorion committed:
Multiple changes

- Renamed "Continuously using" to "start using"
- Refactored (again) the mcitem selector : mcitemall and mcitemblock now have their own code template
- Item consumed and start using item now support both MCitem and ItemCondition
- Removed "(Durability component)" from texts

Goldorion committed:
Remove item_start_using trigger as this is no longer a mandatory minimal feature to have in this PR

It will be added again in a later PR

Goldorion committed:
Remove Durability only and Damage only item component blocks

They are not strictly necessary to have a working PR, so they will be re-added in a later PR)

Goldorion committed:
Remove Item components category

This is not a strictly required feature for this PR, so this will be added again in a later PR. This also doesn't make sense to have a new category (a single block doesn't justify a new category I think)

Goldorion committed:
Remove item components category from toolbox xml

Goldorion committed:
Rename Item condition to Item predicate

Goldorion committed:
Convert Item in inventory trigger

I convert this trigger in this PR as no additional block is required and the converter will need to be adapted for each block in all cases.

Goldorion committed:
I forgot to copy the 1.21.1 file

Goldorion committed:
Merge remote-tracking branch 'upstream/master' into item-entry-related-stuff-advancements

KlemenDEV commented on src/test/java/net/mcreator/integration/generator/GTJSONTriggersBlocks.java:
```diff
@@ -115,13 +115,22 @@ public static void runTest(Logger LOG, String generatorName, Random random, Work
 						</block></next></block></xml>
 						""".formatted(testXML);
-				case "MCItem" -> advancement.triggerxml = """
+				case "MCItem", "ItemPredicate" -> advancement.triggerxml = """
```
Ideally you test ItemPredicate with ItemPredicate still

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -3967,6 +3974,7 @@ blockly.warning.place_feature_ghost_blocks=Some features might not be placed due
 blockly.warning.lake_feature_crash=Due to a bug in worldgen code, lake features containing water will cause the game to crash during world generation.
 blockly.warning.state_provider_int_property_warning=Data pack validation will FAIL if the block doesn''t have the given property, or if the value is out of bounds.
 blockly.warning.utility_adv_trigger=Trigger {0} is a utility advancement trigger and should not be used in combination with other triggers.
+blockly.warning.first_component_type_only=Each type of component only generates the first block of the type. The following blocks will be skipped until a new component type is found.
```
It would be neat to see Blockly validating that only one type of each type being present instead of emitting warning in all cases for UX reasons

KlemenDEV commented:
> (without changing this, we would be required to update all 4 triggerd in this PR)

May be less problematic that having unecessary converter that also changes a lot of files. I prefer the other option without converter

KlemenDEV commented:
> (without changing this, we would be required to update all 4 triggerd in this PR)

Based on this, one option is to add no new triggers and do the following in this PR:

* Add item predicate block
* Update existing blocks to support it

If I understand you correctly, in this case, no FV converter is needed

Goldorion commented:
> Update existing blocks to support it
> If I understand you correctly, in this case, no FV converter is needed


That's the problem. If I update blocks to support it, I need to change the arguments of 3 on 4 triggers and this will break old workspaces. I can't update triggers without making a converter. They both come together.

The only one that doesn't need a converter is the Item consumed trigger, because it has no argument other than the item arg.

Goldorion commented:
I'll take the item in inventory I converted yesterday. The existing trigger had the count params that are parameters existing inside the item ptedicate block, so we duplicate those parameters, but we also duplicate the generated code because they both (the params of the trigger and the item predicate block) need to declare the iten ptedicate JSON object as the count parameter is a sub object of the item predicate block.

For the item damaged trigger, it's tbe same thing but with the minecraft:damage object and the item is enchanted is also the same but with the enchantment parameter.

Goldorion commented:
Pictures with the item damaged trigger:

<img width="667" height="451" alt="image" src="https://github.com/user-attachments/assets/5eee5dba-52e1-44f7-9aac-1882c10ac33e" />

On this picture,m you can see an `item` object inside the conditions object. This item object is what is called an itemPredicate. This ItemPredicate contains different parameters and is exactly what I'm trying to make as an independent block usable by all triggers (supporting ItemPredicate). However, as you can see the existing trigger already declare an item predicate object in order to use the damage parameters, to make work the trigger for users, but I'm also trying to make an output block that encapsulate all this `item` object (so the entire item object is generated with a separate block (the `item_predicate.json` block). We can not keep the old parameters because they are duplicated.

KlemenDEV commented:
I see, thanks for the explanation :D So basically, existing parameters will be moved to the item predicate parameter and if regular item predicate will be used, it will be triggered as soon as item is in the inventory, and that's it.

In this case best would be that this PR does no conversions (and changes no existing blkocks) and just adds item predicate and one new trigger using this predicate or any other item input. It also handles tests and validation of item predicate components.

Then second PR converts existing advancements in single FV number change (and converter) and also changes toolbox presets to use new item predicate.

Did I get it right and my suggestion makes sense?

Goldorion committed:
Revert "I forgot to copy the 1.21.1 file"

This reverts commit a0852b1a8c36b6d5f075a645a96964a90606c35c.

Goldorion committed:
Revert "Convert Item in inventory trigger"

This reverts commit aa3542c30f18f845ad7b47cc0f54f0606f45a1c6.

Goldorion committed:
Revert removing item_start_using

Goldorion committed:
Merge branch 'item-entry-related-stuff-advancements' of https://github.com/Goldorion/MCreator into item-entry-related-stuff-advancements

Goldorion committed:
Some fixes

Goldorion committed:
Workaround to fix tests by allowing min/max fields to be min1/max1

Goldorion committed:
Remove warning (not useful for now)

KlemenDEV commented on plugins/mcreator-core/jsontriggers/item_predicate.json:
```diff
@@ -0,0 +1,48 @@
+  "inputsInline": false,
+  "output": "ItemPredicate",
+  "colour": 90,
+  "extensions": [
```
How do you make sure that it is only generated once?

KlemenDEV replied:
> I think the JSON code generator of MCreator already checks if JSON objects of a JSON object are only generated once. I had nothing to code for this.

It does not, it would help if you understood how and why this works like that

KlemenDEV replied:
Yeah I get that, but we need to understand why this happens. This is undefined behaviour at this time and I expect you as author understand why this happens. Maybe you discovered a bug because this was never intended to work this way

KlemenDEV replied:
Please using debugger or debug prints try to locate where two repeating inputs are reduced into one. I have tried using e.g. feature mod element and there if two same types of blockly blocks are placed, this does not happen. Thanks!

KlemenDEV replied:
Yeah, it makes sense as otherwise this is invalid JSON. I don't think one can disable this and this is normal. Just that we found the reason and understand it :)

If possible I would still try to make validator here in JS as it could be confusing for the user. If you need help, maybe SomeoneElse can help you a bit with writing it, but I think it would not be a complex one

Goldorion replied:
> but I think it would not be a complex one

Everything JS is complex for me 😅😭 And it's even more complex when Intellij doesn't have a JS structure thing like for Java helping you to know how to write the code and what you you can write with methods and variables (I use Community, not Ultimate, which is the one having the JS syntax plugin), so if SomeoneElse can help, I'll appreciate a lot!

KlemenDEV replied:
Please also keep as draft if this needs some more work

KlemenDEV commented on plugins/mcreator-core/blockly/js/mcreator_extensions.js:
```diff
@@ -127,6 +127,36 @@ Blockly.Extensions.register('entity_data_string_list_provider',
+        var minField = this.getField('min');
+        var maxField = this.getField('max');
+
+        if (minField == null) {
```
This is quite unclean, revert this change and merge master where I provided the actual fix. In the future, please take time to try and find correct solutions yourself instead of such workarounds, even if it requires another PR before your main PR idea.

KlemenDEV commented on plugins/mcreator-core/jsontriggers/item_predicate_damage_full.json:
```diff
@@ -0,0 +1,49 @@
+  ],
+  "mcreator": {
+    "group": "damage",
+    "toolbox_id": "components",
```
It is quite hard to know what can be put together and what not from this colorful toolbox, I struggled until I got it right:

<img width="1130" height="747" alt="image" src="https://github.com/user-attachments/assets/86edc94b-0f1e-4b39-9c8d-2197be210ceb" />

I advise you to move item comeponents (and item predicate) stuff into another toolbox or sub-toolbox. You may also want to attach a predicate by default to let user know what should be connected there.

KlemenDEV reviewed (changes requested):
Left some comments. Earlier today you have closed/resolved some of them yourself, please don't do that (this is a rule here for a long time now) as it makes it very hard for me to keep track of things... Thanks for understanding.

<img width="1006" height="274" alt="image" src="https://github.com/user-attachments/assets/2943a5c0-1c5a-47f6-9653-c26d7da90358" />

Goldorion commented on plugins/generator-1.21.8/datapack-1.21.8/jsontriggers/item_predicate_damage_full.json.ftl:
```diff
@@ -0,0 +1,10 @@
+		"min": ${field$min},
+    	"max": ${field$max}
+	},
+	"durability": {
```
As the current item damaged trigger uses the durability object, I would say this is useful (even though the item damaged triggerd doesn't use the correct way, but that's another topic). However, I can see your point. I could simply implement the damage and the durability object in separate blocks and don't allow to define both objects

<img width="621" height="457" alt="image" src="https://github.com/user-attachments/assets/ae6ef4c1-954e-47dc-8a71-cccb8ce2c345" />

KlemenDEV replied:
>  I would say this is useful (even though the item damaged triggerd doesn't use the correct way, but that's another topic)

Yes, it seems like a bug, since durability predicate is "full damage" and damage is "current damage".

Checking predicate by full damage that is fixed property does not actually make sense and you correctly pointed out in the bug we seemed to have in MCreator for some time now :D

Goldorion replied:
What I meant is the trigger contains another object called `delta` and is the object checking for the change in durability. This is the one that would make the more sense to use as like you said the damage->durability cobject check for the full durability while the block's text is more similar to what the delta object does.

<img width="1127" height="523" alt="image" src="https://github.com/user-attachments/assets/9d4718dd-b2b2-4364-8470-b1d0eb462ae8" />

Do you want me to open a PR or is it a too breaking change and the text should be changed?

KlemenDEV replied:
I guess you can just change text in that block for now, but still do it in another PR

Goldorion committed:
Disable duplicate blocks

Goldorion committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into item-entry-related-stuff-advancements

Goldorion committed:
Multiple changes

- Replaced the full damage/durability block by the damage only block
- Item predicate now have the damage component block by default
- Renamed the _mcitemall code template
-

Goldorion commented:
I think I changed everything requested. If there's something missing, then sorry. There was a bit too much big requests at the same time 😅

Goldorion committed:
Tabs

Goldorion committed:
Hardcode values to check for the verification to check all possible cases

Goldorion committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into item-entry-related-stuff-advancements

Goldorion committed:
Change  verification to use a specific format

KlemenDEV commented:
RFR?

KlemenDEV committed:
Merge branch 'master' into item-entry-related-stuff-advancements

KlemenDEV commented on plugins/mcreator-core/blockly/js/mcreator_extensions.js:
```diff
@@ -282,6 +282,38 @@ function checkIfAfter(block, predicate) {
     return null;
 }
 
+Blockly.Extensions.registerMixin('check_duplicate_input_blocks',
```
Validator/extension name should now be something more specific, mentioning type

Goldorion replied:
I renamed it to `disable_duplicate_input_type`. Is it better?

Goldorion committed:
Rename extension +fix the remaining part of something I tested

KlemenDEV committed:
Merge remote-tracking branch 'github-desktop-Goldorion/item-entry-related-stuff-advancements' into item-entry-related-stuff-advancements

KlemenDEV committed:
Merge branch 'master' into item-entry-related-stuff-advancements

KlemenDEV commented on plugins/generator-1.21.1/datapack-1.21.1/jsontriggers/item_start_using.json.ftl:
```diff
@@ -0,0 +1,8 @@
+  "trigger": "minecraft:using_item",
+  "conditions": {
+    "item": {
+    	${input$item}
```
You have tab here, please, go through changed templates again and check if any more tabs are present like here

KlemenDEV reviewed (dismissed):
Found more uses of tabs

KlemenDEV commented on plugins/mcreator-core/blockly/js/mcreator_extensions.js:
```diff
@@ -282,6 +282,38 @@ function checkIfAfter(block, predicate) {
+
+                if (types.has(realType)) {
+                    if (!this.isInFlyout) {
+                        const group = Blockly.Events.getGroup();
```
It would be good to show some sort of warning for UX, similar to e.g. null_comparison_exclude_primitive_types

Goldorion replied:
I tried adding a warning, but the message never appears and I don't know why. There's already a message automatically added when the block is disabled, so I think it's ok (even though the message doesn't specify the specific reason, but the greyed out block tells the user to remove it).
<img width="517" height="90" alt="image" src="https://github.com/user-attachments/assets/213dbc5b-4b10-4b9a-a2be-943ff08b8e10" />

Goldorion replied:
Ok I found the real problem. I simply can not use parameters for the translation inside the JS file.

Goldorion commented on plugins/mcreator-core/blockly/js/mcreator_extensions.js:
```diff
@@ -282,6 +282,38 @@ function checkIfAfter(block, predicate) {
+                return;
+            }
+
+            var types = new Set(); // Store the type of blocks that are already placed in a previous argument.
```
Is there a difference between const and let and one that is more appropriate than the other?

Goldorion replied:
I changed them for const

Goldorion commented on plugins/mcreator-core/jsontriggers/item_predicate_damage_only.json:
It should however be `item_predicate_damage_durability` to follow the format we decided for the extension.

And you're right. I still don't know if I'll add it. This will mainly depend on one trigger.

KlemenDEV replied:
You are right, but this one could still be _damage possibly

Goldorion replied:
Done

KlemenDEV commented on plugins/mcreator-core/blockly/js/mcreator_extensions.js:
@AlsoSomeoneElse, could you check this file when you have time just to double check this is a good approach at this problem? Thanks!

Goldorion committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into item-entry-related-stuff-advancements

Goldorion committed:
Rename damage component file

Goldorion committed:
Tab -> 2 spaces

Goldorion commented on plugins/generator-1.21.1/datapack-1.21.1/jsontriggers/item_predicate.json.ftl:
```diff
@@ -0,0 +1,11 @@
+}</#if><#if input_list$predicateComponent?has_content>,
+"predicates": {
+	<#list input_list$predicateComponent as comp>
+		${comp}
```
Done (for all changed JSON files)

Goldorion committed:
var -> const

Goldorion committed:
Added a warning text when duplicated components

Goldorion committed:
Fix disabled blocks never enabled again

KlemenDEV committed:
Merge remote-tracking branch 'github-desktop-Goldorion/item-entry-related-stuff-advancements' into item-entry-related-stuff-advancements

KlemenDEV committed:
Minor text change

KlemenDEV committed:
Minor format change

Goldorion committed:
Remove stack condition check

Goldorion commented on plugins/mcreator-core/blockly/js/mcreator_extensions.js:
```diff
@@ -282,6 +282,41 @@ function checkIfAfter(block, predicate) {
+                } else {
+                    types.add(realType);
+                }
+                if (!block.isInFlyout) {
```
Fixed 🙂

<img width="1272" height="355" alt="image" src="https://github.com/user-attachments/assets/17a4c3a6-2d19-4ba5-8214-691fd1b3c514" />

Goldorion committed:
Fix extension with multiple types not working in some cases and optimize the code a little bit

Goldorion commented on plugins/generator-1.21.8/datapack-1.21.8/jsontriggers/item_predicate.json.ftl:
```diff
@@ -0,0 +1,11 @@
+${input$item}<#if field$min?number gt 1 || field$max?number != 64>,
```
Fixed

Goldorion commented on plugins/mcreator-core/blockly/js/mcreator_extensions.js:
```diff
@@ -282,6 +282,44 @@ function checkIfAfter(block, predicate) {
+                if (isValid) {
+                    types.add(realType);
+                } else {
+                    warningText = javabridge.t("blockly.extension.disable_duplicate_input_type");
```
Oh ok, sorry

Goldorion committed:
Move javabridge outside of the loop

KlemenDEV commented on plugins/mcreator-core/blockly/js/mcreator_extensions.js:
```diff
@@ -282,6 +282,44 @@ function checkIfAfter(block, predicate) {
+                if (!block.isInFlyout) {
+                    this.setWarningText();
+                    // Makes it so the move and the disable event get undone together.
+                    Blockly.Events.setGroup(e.group);
```
Also you can make disabled and enabled detached blocks which is a bit confusing:

<img width="1897" height="631" alt="image" src="https://github.com/user-attachments/assets/da7bec1a-e045-4b83-9751-b5b22c330130" />

not sure if this is easily fixable, though

KlemenDEV replied:
@Goldorion, this was fixed. Please see the code so you know how to do it next time. The fix was mostly just moving this to child block and then looking its parent, so child validates itself.

Goldorion commented on plugins/generator-1.21.8/datapack-1.21.8/jsontriggers/item_start_using.json.ftl:
I tested with a procedure printing the itemstack's display name in the chat with the Player starts using item global trigger (I locked the code as it doesn't exist inside MCreator) and when I was inside the game, eating an apple sent the apple's display name, but using a pickaxe wasn't sending the pickaxe's display name inside the chat. This means that pickaxes (and other tools I guess) are not supported by this trigger...

This also means we should change the name of the trigger as it may confuse users. The Wiki says this about this trigger:
<img width="1026" height="160" alt="image" src="https://github.com/user-attachments/assets/83bd58d9-35bc-4cea-93ba-1ea0380aa721" />

KlemenDEV replied:
Indeed, we need to be very clear what this trigger does and what not, otherwise there will be many confused user bug reports

Goldorion committed:
Remove Item start using trigger

Goldorion committed:
Change predicate components to SomeoneElse's format

Goldorion committed:
Revert "Change predicate components to SomeoneElse's format"

This reverts commit 4012a53c79d591b858f0c61c691f8f79fe5d8011.

KlemenDEV committed:
First, let's keep things ordered

KlemenDEV committed:
Merge branch 'master' into item-entry-related-stuff-advancements

KlemenDEV committed:
This should do the trick. Make sure to apply the disable_duplicate_input_type to predicate types when adding new ones

KlemenDEV committed:
Revert unrelated changes

KlemenDEV commented:
> I know you can use the item is consumed quite simply

There is not a single item in the game that you can consume and has damage (without resorting to command tricks). So you need to understand we can't really showcase the feature like that.

> However, if you mean a new trigger, I woukd have to search (but I don't think so).

>> If nothing else, maybe the one you removed is fine, just use better text, such as "If item X with use duration is used"

KlemenDEV committed:
Merge branch 'master' into item-entry-related-stuff-advancements

Goldorion committed:
_mcitemblock -> _mcblock

KlemenDEV commented:
The actual issue I was referring to remains, however:

<img width="1485" height="351" alt="image" src="https://github.com/user-attachments/assets/67a874cd-f299-40df-9272-73edb3e159be" />

KlemenDEV commented on src/test/java/net/mcreator/integration/generator/GTJSONTriggersBlocks.java:
```diff
@@ -40,6 +40,7 @@
 
 public class GTJSONTriggersBlocks {
 
+	@SuppressWarnings("DuplicateBranchesInSwitch")
```
I had a brainfart here, you can indeed merge those since the block you put the ouput block in is the same

Goldorion committed:
Klemen's brainfart

Goldorion committed:
Revert "_mcitemblock -> _mcblock"

This reverts commit 2bfb2c33edf14598cc8b810196e7535596a1dafd.

Goldorion committed:
Name the variable item when this makes sense

Goldorion commented:
I pushed the change

KlemenDEV commented:
This PR is now ok with the code it has, but it lack a usable block as suggested in this comment:

<img width="1442" height="302" alt="image" src="https://github.com/user-attachments/assets/41755844-2c88-403e-a66a-e1ee9b78892a" />

I suggest you add that trigger back, just change wording to be more clear with what items it works

Goldorion committed:
Add (once again) the using_item trigger

AlsoSomeoneElse commented on plugins/mcreator-core/jsontriggers/item_predicate.json:
```diff
@@ -0,0 +1,48 @@
+    {
+      "type": "field_number",
+      "name": "max",
+      "value": 64,
```
I'd set initial value to 99 and add a max value to both number fields.

Goldorion committed:
99 limit + condition

Goldorion committed:
Merge remote-tracking branch 'upstream/master' into item-entry-related-stuff-advancements

KlemenDEV commented:
I guess it makes sense to call them at least internally what they actually are, but I would keep UI texts as they are for now for easier understanding for users. 

Goldorion, if you do the rename, try to contain it in one commit so it is easier to track changes

Goldorion committed:
item_predicate_damage -> data_component_predicate_damage

Goldorion commented:
> I'd also rename `ItemPredicateEntry` output type to `DataComponentPredicate`

This is what I was also thinking, so I did the change inside the same commit :)

AlsoSomeoneElse commented on plugins/generator-1.21.1/datapack-1.21.1/jsontriggers/_mcitem.json.ftl:
```diff
@@ -0,0 +1,2 @@
+<#include "mcitems.ftl">
```
This should probably be `mcitems_json.ftl`, following fd3aa311d09b1d7585e0c0a52bdcdad9a10ff8a8

Goldorion committed:
Merge remote-tracking branch 'upstream/master' into item-entry-related-stuff-advancements

Goldorion committed:
mcitems_json

Goldorion committed:
Merge remote-tracking branch 'upstream/master' into item-entry-related-stuff-advancements

Goldorion committed:
Merge remote-tracking branch 'upstream/master' into item-entry-related-stuff-advancements

KlemenDEV commented:
Following the merge of 26.1.x, this PR should merge master into it and add support for 26.1.x.

You can also leave as draft and wait until 1st 2026.2 snapshot, when 1.21.8 will be removed to do both at the same time

KlemenDEV commented:
Now that 1.21.8 was removed, delete its support in the PR too

Goldorion committed:
Merge remote-tracking branch 'upstream/master' into item-entry-related-stuff-advancements

Goldorion committed:
Merge remote-tracking branch 'upstream/master' into item-entry-related-stuff-advancements

# Conflicts:
#	plugins/generator-1.21.8/datapack-1.21.8/jsontriggers/any_item_in.json.ftl
#	plugins/generator-1.21.8/datapack-1.21.8/jsontriggers/item_consumed.json.ftl
#	plugins/generator-1.21.8/datapack-1.21.8/jsontriggers/item_damaged.json.ftl
#	plugins/generator-1.21.8/datapack-1.21.8/jsontriggers/item_enchanted.json.ftl
#	plugins/generator-1.21.8/datapack-1.21.8/jsontriggers/item_in_inventory.json.ftl

Goldorion committed:
Add 26.1.x + Remove 1.21.8

Goldorion committed:
I pushed the previous by accident, so here's the commit fixing things after a round of tests

Goldorion committed:
I should maybe remove 1.21.8 😅

Goldorion committed:
Apparently, there was one more error to fix

KlemenDEV commented:
Were all JS features carefully tested following Blockly update?

Goldorion commented:
Yes

KlemenDEV commented:
There is a minor merge conflict

Goldorion committed:
Merge remote-tracking branch 'upstream/master' into item-entry-related-stuff-advancements

# Conflicts:
#	plugins/mcreator-core/blockly/js/mcreator_extensions.js

KlemenDEV reviewed (approved):
LGTM, thanks!
